Sanity Media Plugin

diboune opened the pull request.

diboune committed:
media plugin base

diboune committed:
Add media library with bulk upload, filtering, browser dialog, and selection mode

Features:
- Bulk upload with staging dialog and metadata fields
- Filtering, sorting, search, and pagination
- Media browser dialog for asset selection
- Selection mode with context management
- Grid and list views for media assets
- Video input improvements

Acked-by: Human

diboune committed:
quick fixes

stilyan-tinloof commented on examples/blog-next/package.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@examples/blog-next",
-	"version": "0.1.7",
+	"version": "0.0.0-20260112150832",
```
Let's change the version back to 0.1.7

stilyan-tinloof commented on examples/blog-studio/src/schemas/objects/sections/image.ts:
```diff
@@ -1,21 +1,44 @@
-			title: "image.alt",
-		},
-	},
+  name: "section.image",
```
Do we still need this if we have media-image? Or do we have it to show off the differences?

stilyan-tinloof commented on examples/blog-studio/src/schemas/objects/sections/media-image.ts:
```diff
@@ -0,0 +1,33 @@
+	],
+	preview: {
+		select: {
+			alt: "image.alt",
```
`image.alt` is returning nothing, should be `image.asset.alt`

stilyan-tinloof commented on examples/blog-studio/package.json:
```diff
@@ -1,7 +1,7 @@
 	"name": "@examples/blog-studio",
 	"private": true,
-	"version": "0.0.6",
+	"version": "0.0.0-20260112150832",
```
Let's revert the version number back to 0.0.6

stilyan-tinloof commented on packages/sanity-media/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2024 Tinloof
```
Bro's living in the past

stilyan-tinloof commented on packages/sanity-media/src/components/media-panel/components/delete-confirm-dialog.tsx:
```diff
@@ -0,0 +1,66 @@
+  return (
+    <Dialog
+      id="delete-dialog"
+      header="Delete asset"
```
Check for whether it is bulk and add an s

stilyan-tinloof commented on packages/sanity-media/src/components/shared/hooks/use-bulk-selection.ts:
```diff
@@ -0,0 +1,225 @@
+        try {
+          // Delete thumbnail if it's a video
+          if (singleAsset.mediaType === "video" && singleAsset.thumbnail?._id) {
+            await client.delete(singleAsset.thumbnail._id);
```
For all the other bulk operations we are using transactions but not for deleting

stilyan-tinloof commented on packages/sanity-media/src/types.ts:
```diff
@@ -0,0 +1,59 @@
+  description?: string;
+}
+
+export function isVideoContentType(contentType: string) {
```
Functions inside of a types.ts file

stilyan-tinloof reviewed (approved):
Looks good, nothing breaking at all

Also don't forget a changeset

I've also created a Linear issue of [TIN-3546: Future improvements for `@tinloof/sanity-media`](https://linear.app/tinloof/issue/TIN-3546/future-improvements-for-tinloofsanity-media)

stilyan-tinloof committed:
Fix media browse navigation returning to wrong tool

Store the full source URL when entering selection mode from an input
field instead of relying on Sanity's intent URL system. This ensures
users return to the exact same tool they were in (e.g., structure tool)
rather than whichever tool is first in the plugins config.

stilyan-tinloof committed:
Use tri-state checkboxes for metadata filters

Replace separate "has X" and "missing X" checkboxes with a single
tri-state checkbox for each metadata field (alt, title, caption):
- Unchecked: No filter applied
- Checked: Filter for items that have the metadata
- Indeterminate: Filter for items missing the metadata

Clicking cycles through all three states.

stilyan-tinloof committed:
Address PR review comments

- Revert version/changelog files for blog-next, blog-studio, sanity-media
- Fix schema issues: use image.asset.alt and video.asset.title, remove subtitle/caption
- Remove unused imports in upload-handler.ts
- Add plural 's' for bulk delete dialog header
- Create API_VERSION constant and use consistently across all files
- Fix onBlur firing without changes in media-detail-panel.tsx
- Refactor types.tsx to types.ts, extract DebouncedSearchInput component
- Create constants.ts for shared constants (PAGE_SIZE, SORT_OPTIONS, etc.)
- Create utils.ts for utility functions (formatDuration, formatFileSize, isVideoContentType, isImageContentType)
- Use transactions for bulk delete in use-bulk-selection.ts
- Add cleanup for setTimeout in use-upload-queue.ts
- Remove unused maxFileSize from MediaStoragePluginConfig
- Update LICENSE year to 2026

stilyan-tinloof committed:
lockfile

stilyan-tinloof committed:
Auto-select uploaded asset and return to field

- Capture asset ID from upload handler result
- Auto-select first successful upload and close dialog
- Stay in upload tab only if all uploads failed

stilyan-tinloof committed:
Improve media browser dialog: simplify UX and fix filters

- Remove upload tab, keep upload button with single-file auto-select
- Remove footer with Select/Cancel buttons, click asset to select
- Fix counts and usage counts to respect assetType filter
- Exclude draft documents from counts
- Reduce grid columns for better card sizing
- Fix double scrollbar issue with proper flex layout
- Add collapsible filter UI with active filter chips
- Use segmented control for usage filter
- Use tri-state checkboxes for metadata filters with dynamic labels
- Remove color indicators from tag pills
- Open browse dialog instead of navigating to Media tool

stilyan-tinloof committed:
Refactor media tool UI: remove tabs, add settings button, sticky header

- Remove media/settings tabs from media-tool.tsx
- Add settings button (cog icon) next to upload button
- Add "Back to Media" button in settings panel header
- Make media library title, upload, and filters sticky
- Separate sidebar toggle from view toggle buttons
- Use tri-state checkboxes for metadata filters with dynamic labels
- Remove color indicators from tag pills

stilyan-tinloof committed:
Add dynamic tag width calculation and sync grid components

- Create TagList component that measures available width and shows only tags that fit
- Use ResizeObserver to recalculate on resize
- Show +N indicator for hidden tags with accurate count
- Make type indicator same size as checkbox (14px)
- Sync asset-grid.tsx and media-grid-view.tsx styling

stilyan-tinloof committed:
Redesign media sidebar and standardize UI components

- Standardize button padding/fontSize across all components
- Redesign sidebar: filter icon, toggle edit mode for tags, usage toggle group
- Style metadata filters as colored buttons (green=has, amber=missing)
- Add Sanity Dialog for tag deletion instead of browser confirm
- Move "Clear all" filters button to sidebar header
- Remove filter chips display from main content area
- Add borders between sidebar sections

stilyan-tinloof committed:
Fix asset deletion to also remove files from storage

- Add deleteFile function to storage-client using S3 DeleteObjectCommand
- Update MediaAsset type to include path field
- Update GROQ query to fetch path for assets and thumbnails
- Update useBulkSelection to delete files from storage before removing Sanity documents

stilyan-tinloof committed:
Only show settings button in development mode

stilyan-tinloof committed:
Add hasAudio field to video assets

stilyan-tinloof committed:
Fix TypeScript error for audioTracks property

stilyan-tinloof committed:
Fix Safari video thumbnail and audio detection issues

- Force video playback to decode frames on Safari
- Add blank canvas detection and retry logic
- Improve audio track detection after playback
- Handle autoplay blocking gracefully

stilyan-tinloof committed:
Add manual hasAudio toggle in video detail panel

- Add hasAudio field to MediaAsset metadata type
- Add Switch component to toggle hasAudio on/off
- Update media asset via patch when toggled
- Default to true if hasAudio is not set

stilyan-tinloof committed:
Improve detail panel styling and consistency

- Add border around Details and References tabs
- Change all labels to sentence case (File info, Alt text, Has audio, etc.)
- Standardize filename layout to match other fields
- Use × symbol for dimensions instead of x
- Update Media Assets to Media assets

stilyan-tinloof committed:
fix(sanity-media): use adapter typePrefix for tag type to avoid naming conflicts

The tag type now uses the adapter's typePrefix (e.g., r2.tag instead of media.tag)
to avoid conflicts with sanity-plugin-media which also defines media.tag.

- Update generateTagType() to accept adapter and use ${typePrefix}.tag
- Update tag references in asset type generators
- Update useTags and useTagEditor hooks to accept adapter parameter
- Update all call sites to pass adapter

stilyan-tinloof committed:
feat(sanity-media): add customizable tool name/title and settings button

- Add toolName and toolTitle options to MediaPluginOptions for avoiding conflicts
- Add "Open Settings" button on credentials warning (dev mode only)

stilyan-tinloof committed:
docs(sanity-media): add architecture documentation

stilyan-tinloof committed:
feat(sanity-media): add upload staging dialog to input fields

- Show staging dialog before upload to allow adding metadata (alt, caption, title, description) and tags
- Extract createStagingItem helper to reduce duplication
- Add credentials check to handleInputChange for consistency
- Fix loose types (assetPreview, handleBrowseSelect use MediaAsset)
- Remove unused Box import and dead handleFileSelect function

stilyan-tinloof committed:
feat(sanity-media): add inline metadata editing to input fields

- Add alt text and caption fields to image input
- Add title, description, and has audio toggle to video input
- Update detail panel styling to match native Sanity form fields
- Move has audio toggle from file info to metadata section in detail panel
- Fix type error in loadAsset function

stilyan-tinloof committed:
fix(sanity-media): hide internal document types from create menu

Filter out tag, imageAsset, fileAsset, and videoAsset document types
from the global "Create new document" menu using newDocumentOptions.

stilyan-tinloof committed:
feat(sanity-media): add drag-and-drop file upload to media panel

Add drag-and-drop support for uploading files directly to the media panel.
Files dragged onto the panel open the staging dialog for metadata editing
before upload. Includes a visual drop zone overlay with blur effect.

stilyan-tinloof committed:
fix(sanity-media): make asset cards focusable for paste events

Add tabIndex={0} to the "has asset" Card components in both
MediaImageInput and MediaVideoInput. This ensures the cards can
receive keyboard focus, allowing paste events to work correctly.

stilyan-tinloof committed:
feat(sanity-media): improve upload staging dialog UX

- Redesign file cards with thumbnails, better spacing and layout
- Add sticky footer with action buttons (always visible)
- Add keyboard navigation (Arrow Up/Down, j/k to navigate)
- Add Enter/Space to expand/collapse items with auto-focus on first input
- Add Delete/Backspace to remove focused item
- Add Escape to blur input or close dialog
- Add Cmd/Ctrl+Enter to start upload
- Smooth scroll to keep focused item visible
- Show missing alt text count in upload button
- Match input field styles with native Sanity components

stilyan-tinloof committed:
fix(sanity-media): use native Button component for Upload in inputs

Replace custom styled UploadButton label with Sanity UI Button component
to match the Browse button styling (font, padding, mode).

stilyan-tinloof committed:
fix(sanity-media): center loading spinner in image/video inputs

- Use absolutely positioned Flex to center spinner during asset loading
- Hide media with opacity while loading to prevent layout shift
- Remove unused LoadingOverlay styled components
- Add loading state for when asset ref exists but preview hasn't loaded

stilyan-tinloof committed:
chore(sanity-media): update dependencies to match other sanity packages

- Update react/react-dom to ^19.2.4
- Update sanity to ^5.7.0
- Update styled-components to ^6.3.8
- Update @types/react to ^19.2.10
- Make peerDependencies more specific

stilyan-tinloof committed:
feat(sanity-media): generate optimized preview thumbnails for media browsing

Generate WebP preview thumbnails (max 600px height) during image upload
to improve media panel performance when browsing large images.

- Add generatePreviewBlob() for creating optimized thumbnails
- Add uploadBlob() and getPreviewKey() storage helpers
- Store preview URL in asset documents
- Update grid/list views to use preview with fallback chain
- Extract ThumbnailCell component and getAssetPreviewUrl helper
- Clean up preview files on asset deletion
- Handle CORS failures gracefully with fallback
- Classify errors in preview deletion (only ignore 404s)

stilyan-tinloof committed:
feat(sanity-media): use container queries for responsive grids and improve media browser dialog layout

- Add ResponsiveGrid component using CSS container queries for container-based responsiveness
- Replace viewport-based Grid columns with container-query-based grid in media-grid-view, asset-grid, and media-browser-dialog
- Restructure media browser dialog with proper section separation (header, content, pagination footer)
- Add border dividers between header/content and content/pagination sections
- Fix scroll behavior so cards scroll under borders seamlessly
- Reorganize toolbar into two rows for better small container layout

stilyan-tinloof committed:
feat(sanity-media): add sticky pagination footer and item count to media panel heading

- Move pagination to sticky footer matching media browser dialog layout
- Display item count next to "Media Library" heading as "(X items)"
- Remove redundant item count from scrollable content area

stilyan-tinloof committed:
feat(sanity-media): add enhanced pagination component with page numbers

- Add first/last page buttons using double chevron icons
- Show clickable page numbers with ellipsis for large page counts
- Current page highlighted with background, others muted
- Accessible buttons with proper keyboard focus
- Replace simple "Page X of Y" with interactive pagination in media panel and dialog

stilyan-tinloof committed:
fix(sanity-media): scroll to top on page change in media browser dialog

stilyan-tinloof committed:
feat(sanity-media): show upload progress in staging dialog with auto-close

- Move upload progress display from media panel to staging dialog
- Show per-item progress bars with status icons during upload
- Auto-close dialog 3 seconds after all uploads complete
- Remove separate upload queue in favor of tracking state in staging items

stilyan-tinloof committed:
feat(sanity-media): add useMediaPasteHandler hook for Portable Text

- Add hook to intercept image/video pastes in PortableTextInput
- Support multiple files in a single paste
- Show staging dialog for metadata entry before upload
- Return all uploaded assets in batched callback
- Auto-close dialog after uploads complete

stilyan-tinloof committed:
fix(sanity-media): address code review issues in upload hooks

- Fix memory leak: use functional state update in closeStagingDialog
- Add unmount cleanup effect to revoke object URLs
- Fix ID collision: add counter to prevent duplicate IDs when multiple
  files are added/pasted simultaneously
- Make expanded state consistent: only first item expanded in both hooks
- Add missing 'uploading' dependency to renderStagingDialog

stilyan-tinloof committed:
fix(sanity-media): address high priority code review issues

- Fix incorrect useState hook usage in media-file-input.tsx by
  changing it to useEffect for side effects
- Replace window.location.href with router.navigateUrl in
  selection-context.tsx to preserve unsaved state and avoid
  full page reloads
- Fix race condition in pending selection by clearing session
  storage immediately and removing fragile ref-based guard

stilyan-tinloof committed:
fix(sanity-media): address medium and low priority code review issues

- Fix memory leak in metadata-extractor by ensuring Object URLs are
  revoked in all error paths via try-catch-finally
- Add GROQ query escaping to prevent special character issues and
  limit search length to 100 chars
- Add filename length validation (200 char limit) in storage-client
  to prevent S3 key limit issues
- Add toast notifications for metadata update errors with automatic
  state revert on failure
- Document LQIP_MAX_DIMENSION constant explaining 20px tradeoff
- Replace any type with FileAssetPreview type in media-file-input

stilyan-tinloof committed:
feat(sanity-media): improve grid checkbox UX with responsive visibility

- Resize checkboxes and type indicators to 20px (from 14px)
- Desktop: hide checkboxes by default, show on hover/focus/selected
- Mobile/touch: always show checkboxes for accessibility
- Add smooth opacity transitions

stilyan-tinloof committed:
feat(sanity-media): redesign media panel header with responsive layout

- Reorganize header into two rows:
  - Row 1: Full-width search + Upload + Settings buttons
  - Row 2: Select all checkbox + Title/count + Type filter + Sort + View toggle
- Make toolbar responsive on mobile:
  - Hide button text for type/sort filters (show only icons)
  - Hide "Media Library" title, keep item count visible
- Improve selection UX:
  - Show "X of Y selected" instead of replacing title
  - Add accessible select all checkbox with keyboard support
  - Remove bulk delete from toolbar (user has another plan for this)
- Simplify view toggle to single button with tooltip
- Add icons to sort menu items (Calendar, DocumentText, Database)

Note: CSS selectors for hiding button text use internal data-ui attributes
which may need updates if Sanity UI changes its internal structure.

stilyan-tinloof committed:
feat(sanity-media): improve select all checkbox with indeterminate state

- Show checkmark when all items selected, minus icon when partial
- Simplify count text to always show total items
- Change asset checkboxes from CheckmarkCircleIcon to CheckmarkIcon

stilyan-tinloof committed:
revert(sanity-media): remove 600px preview thumbnail feature

This removes the optimized preview thumbnail generation that was added
in commit 1029498. Preview thumbnails are no longer generated or stored
in Cloudflare R2 storage.

Changes:
- Remove preview generation from upload handler
- Remove getPreviewKey and uploadBlob from storage client
- Remove preview constants (PREVIEW_MAX_HEIGHT, PREVIEW_QUALITY, PREVIEW_SUFFIX)
- Remove generatePreviewBlob from metadata extractor
- Remove preview field from imageAsset and videoAsset schemas
- Update getAssetPreviewUrl to use video thumbnails directly
- Clean up preview deletion code from bulk selection hook

stilyan-tinloof committed:
fix(sanity-media): prevent flash of loading states in media panel

Fixed race conditions causing brief flashes of incorrect UI states during initial load:

- Add intermediate state detection for credentials loading to prevent config message flash
- Track countsLoading state to prevent premature "No media yet" message
- Add debug console logs to monitor render state transitions

The component now properly handles the gap between credentialsLoading becoming false and credentials being set by detecting when media exists but credentials aren't ready yet.

stilyan-tinloof committed:
refactor(sanity-media): fix loading state race condition at the source

Fixed the root cause of loading state flashes by ensuring useCredentials hook keeps loading state true until credentials are fully processed, not just when the API call completes.

Changes:
- Add credentialsReady state to track when credentials are fully set
- Keep loading=true until both secretsLoading is done AND credentials are processed
- Handle edge cases: conversion errors and missing secrets
- Remove debug console logs from media panel
- Remove intermediate state detection workaround (no longer needed)

This eliminates the 1-render gap that was causing UI flashes during initial load.

stilyan-tinloof committed:
fix(sanity-media): harden useCredentials against edge cases

Address code review findings to make the hook more robust:

- Add adapter memoization by id to prevent infinite re-renders from reference changes
- Reset state when adapter or secrets change to prevent stale credentials
- Clear credentials on conversion errors instead of keeping stale values
- Add effect cleanup with cancellation flag to prevent race conditions
- Add JSDoc for credentialsReady state

These fixes handle edge cases that could occur during adapter changes, rapid secrets updates, or conversion errors.

stilyan-tinloof committed:
fix(sanity-media): remove broken memoization and unnecessary state resets

Simplify useCredentials hook based on code review findings:

- Remove broken useMemo that could return stale adapter functions
- Remove state reset at effect start that caused 3-render cycle
- Use adapter directly since it's stable at plugin initialization
- Add comprehensive JSDoc for return values
- Keep effect cleanup and error handling improvements

This results in cleaner, more correct code without extra re-renders.

stilyan-tinloof committed:
feat(sanity-media): add imageTransformer option for CDN image optimization

Adds an optional imageTransformer function to MediaPluginOptions that allows
consumers to route image URLs through a CDN (e.g. Cloudflare Image Resizing,
imgix) without any CDN-specific logic in the plugin itself.

- Adds ImageTransformer type and ImageTransformerOptions type (exported from public API)
- Threads imageTransformer through AdapterContext for input components
- Props-drills imageTransformer through MediaTool → MediaPanel → grid/list/detail views
- Applied to images only (never video thumbnails): ThumbnailCell (200×200),
  AssetGrid/MediaGridView (450×450), MediaDetailPanel (900 scale-down),
  MediaImageInput (600 cover)
- Sizes align with main site srcset widths for shared Cloudflare cache entries
- Fully optional: plugin behaviour is unchanged when not provided

stilyan-tinloof committed:
refactor(sanity-media): co-locate ImageTransformer types, enable video thumbnail transformation

- Move ImageTransformer type into types.ts alongside ImageTransformerOptions so
  both are co-located and exported from a single source in the public API
- Remove unused quality field from ImageTransformerOptions (was defined but never
  passed at any call site — avoids false promises in the public API)
- Enable video thumbnail transformation: remove mediaType === "image" guard from
  ThumbnailCell, AssetGrid, and MediaGridView — getAssetPreviewUrl always returns
  an image URL (the asset itself for images, the thumbnail image for videos), so
  both benefit equally from CDN optimisation
- Strengthen MediaImageInput guard to use _type.endsWith("videoAsset") instead of
  the derived mediaType field which is not stored on the Sanity document

stilyan-tinloof committed:
feat(sanity-media): add presigned URL upload support

Extend StorageAdapter with optional presign/presignDelete functions so
consumers can use server-side presigned URLs instead of browser-side
AWS credentials. Existing credential-based adapters remain unchanged.

A centralized `ready` boolean replaces scattered `!credentials` guards
across all input components, the media panel, upload queue, bulk
selection, and browser dialog.

stilyan-tinloof committed:
fix(sanity-media): handle structured title/description objects in GROQ queries

Extract .text from title/description fields that may be objects (e.g. {alignment, text})
instead of plain strings, preventing React "Objects are not valid as a React child" errors.

strlrd-29 committed:
feat(sanity-media): relabel video description field to caption

strlrd-29 committed:
Merge remote-tracking branch 'origin/main' into tin-3485-sanity-media-plugin

# Conflicts:
#	examples/blog-next/package.json
#	examples/blog-studio/package.json
#	pnpm-lock.yaml
